programmazione-concorrente.md

```diff
@@ -1 +1,437 @@
+- float: AtomicInteger con i metodi `Float.floatToIntBits(float)` e `Float.intBitstoFloat(int)`
+- double: AtomicLong con i metodi `Double.doubleToLongBits(double)` e `Double.longBitsToDouble(long)`
+
+oppure, se è necessario sommare, utilizzando `DoubleAdder`:
+```java
+class Account {
+    private final DoubleAdder balance = new DoubleAdder();
+    
+    public float get() { 
+        return (float) balance.sum(); 
+    }
+    
+    public void deposit(float money) { 
+        balance.add(money); 
+    }
+
+    public void withdraw(float money) {
+        balance.add(-money); 
+    }
+}
+```
+
+Particolare attenzione va posta anche nella dichiarazione di array:
+`volatile int[] array = new int[20]` dichiara un array di interi con referenza volatile, _non_ un array di 
+interi volatili. Bisogna quindi utilizzare `AtomicIntegerArray`, `AtomicLongArray` o `AtomicReferenceArray<E>`.
+
+Il vantaggio è che, nonostante quello che dicono le slide (di nuovo), mandare un intero thread in WAIT quando devi 
+semplicemente leggere una variabile dalla memoria centrale è un'idea del cazzo. Semplicemente devi dire al compilatore
+di leggere dalla memoria.
+
+## Variabili finali e oggetti immutabili
+
+Le variabili finali non hanno bisogno di essere sincronizzate in alcuna maniera, in quanto, non potendo cambiare,
+è impossibile che avvenga un desync tra la cache locale e la memoria centrale.
+
+Questo ci porta agli oggetti immutabili: un oggetto immutabile è tale che ogni suo campo è dichiarato `final` e
+ogni modifica dello stesso avviene mediante la creazione di un nuovo oggetto:
+```java
+public class ImmutableRGB {
+    private final String name;
+    // Values must be between 0 and 255.
+    private final int red;
+    private final int green;
+    private final int blue;
+
+    private int check(int component) {
+        if (component < 0 || component > 255)
+            throw new IllegalArgumentException();
+        return component;
+    }
+    
+    public ImmutableRGB(int red, int green, int blue, String name) {
+        this.red = check(red);
+        this.green = check(green);
+        this.blue = check(blue);
+        this.name = name;
+    }
+
+    public int getRGB() {
+        return (red << 16) | (green << 8) | blue;
+    }
+    
+    public String getName() {
+        return name;
+    }
+
+    public ImmutableRGB invert() {
+        return new ImmutableRGB(255 - red, 255 - green, 255 - blue, "Inverse of " + name);
+    }
+}
+```
+
+Gli oggetti immutabili non necessitano di alcun tipo di sincronizzazione.
+
+## Producer/consumer pattern
+
+Java mette a disposizione nell'intrinsic lock anche delle primitive per implementare dei **Guarded Blocks**,
+ovvero delle condizioni che devono essere vere prima che l'esecuzione di un metodo possa continuare:
+- `wait() throws InterruptedException`: sospende l'esecuzione di un thread fino a quando viene risvegliato da un altro
+- `notify()`: sveglia un thread sospeso per riprendere la sua esecuzione
```

Comment:
Sia per `notify` che `notifyAll` specificherei che risveglia i thread in attesa su quel particolare oggetto.
La chiarezza non è mai troppa 😂